sound/soc/sof/intel/hda-dsp.c

```diff
@@ -431,11 +435,62 @@ static int hda_dsp_set_D0_state(struct snd_sof_dev *sdev,
 		    !hda_enable_trace_D0I3_S0 ||
 		    sdev->system_suspend_target != SOF_SUSPEND_NONE)
 			flags = HDA_PM_NO_DMA_TRACE;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+		/*
+		 * Before system suspend CORB/RIRB DMA must be stopped and links powered off.
+		 * But when the system is in S0, this should be done only when there are no
+		 * codecs powered, in order to save power.
+		 */
+		if (sdev->system_suspend_target == SOF_SUSPEND_NONE && bus->codec_powered)
+			goto set_d0i3_reg;
+
+		/* stop the CORB/RIRB DMA if it is On */
+		if (bus->cmd_dma_state)
+			snd_hdac_bus_stop_cmd_io(bus);
+
+		/* power off links */
+		ret = snd_hdac_ext_bus_link_power_down_all(bus);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error %d in %s: failed to power down links",
+				ret, __func__);
+			return ret;
+		}
+#endif
 	} else {
 		/* prevent power gating in D0I0 */
 		flags = HDA_PM_PPG;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+		/*
+		 * During system resume, CORB/RIRB DMA and links must be brought to the same state
+		 * as before suspend.
+		 * But when the system is in S0, this should be done only if the codecs were
+		 * powered off before entering D0i3.
+		 */
+		if (sdev->system_suspend_target == SOF_SUSPEND_NONE && bus->codec_powered)
+			goto set_d0i3_reg;
+
+		/* power up links that were active */
+		list_for_each_entry(hlink, &bus->hlink_list, list) {
+			if (hlink->ref_count) {
+				ret = snd_hdac_ext_bus_link_power_up(hlink);
+				if (ret < 0) {
+					dev_err(sdev->dev,
+						"error %d in %s: failed to power up links",
```

Comment:
format is not consistent and missing \n
should be "error: %s: failed to power up links %d\n"

Reply:
fixed now. I removed the %s and left just the return value.